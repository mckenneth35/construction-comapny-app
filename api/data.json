[
  {
    "id": 1,
    "name": "Albert Werner Excavation",
    "logoUrl": "https://placekitten.com/100/100",
    "specialties": [
      "EXCAVATION"
    ],
    "city": "Berlin"
  },
  {
    "id": 2,
    "name": "Newton Einstein Edison Mozart da Vinchi McDonald Excavation",
    "logoUrl": "https://placekitten.com/200/200",
    "specialties": [
      "EXCAVATION"
    ],
    "city": "London"
  },
  {
    "id": 3,
    "name": "GWS Construction",
    "logoUrl": "https://placekitten.com/200/250",
    "specialties": [
      "EXCAVATION",
      "PLUMBING",
      "ELECTRICAL",
      "FIRE_SAFETY"
    ],
    "city": "Berlin"
  },
  {
    "id": 4,
    "name": "Klusenburg IR Pro",
    "logoUrl": "https://placekitten.com/220/200",
    "specialties": [
      "EXCAVATION",
      "PLUMBING"
    ],
    "city": "Lens"
  },
  {
    "id": 5,
    "name": "Meyer & Partner Construction",
    "logoUrl": "https://placekitten.com/520/530",
    "specialties": [
      "EXCAVATION",
      "ELECTRICAL"    ],
    "city": "Santa Claus Village Joulupukin Pajakyl√§"
  },
  {
    "id": 6,
    "name": "Nother-Bau",
    "logoUrl": "https://placekitten.com/200/400",
    "specialties": [
      "PLUMBING"
    ],
    "city": "Munich"
  },
  {
    "id": 7,
    "name": "Prignitz",
    "logoUrl": "https://placekitten.com/400/400",
    "specialties": [
      "FIRE_SAFETY"
    ],
    "city": "Antwerp"
  },
  {
    "id": 8,
    "name": "Sommer Road Building",
    "logoUrl": "https://placekitten.com/400/200",
    "specialties": [
      "EXCAVATION",
      "PLUMBING",
      "ELECTRICAL"
    ],
    "city": "Geneva"
  },
  {
    "id": 9,
    "name": "Buenos Dias Construct",
    "logoUrl": "https://placekitten.com/500/200",
    "specialties": [
      "EXCAVATION",
      "FIRE_SAFETY"
    ],
    "city": "London"
  },
  {
    "id": 10,
    "name": "Marco Reus Electric",
    "logoUrl": "https://placekitten.com/450/360",
    "specialties": [
      "ELECTRICAL"
    ],
    "city": "Dortmund"
  },
  {
    "id": 11,
    "name": "All Rounder Co",
    "logoUrl": "https://placekitten.com/350/300",
    "specialties": [
      "EXCAVATION",
      "PLUMBING",
      "ELECTRICAL",
      "FIRE_SAFETY"
    ],
    "city": "Paris"
  },
  {
    "id": 12,
    "name": "Musical Construction",
    "logoUrl": "https://placekitten.com/500/700",
    "specialties": [
      "PLUMBING",
      "ELECTRICAL"
    ],
    "city": "Vienna"
  },
  {
    "id": 13,
    "name": "Wenger & Wagner & Werner Excavation",
    "logoUrl": "https://placekitten.com/60/50",
    "specialties": [
      "EXCAVATION",
      "ELECTRICAL"
    ],
    "city": "Berlin"
  },
  {
    "id": 14,
    "name": "Alpha Wenger Excavation",
    "logoUrl": "https://placekitten.com/500/500",
    "specialties": [
      "PLUMBING",
      "ELECTRICAL",
      "FIRE_SAFETY"
    ],
    "city": "Berlin"
  },
  {
    "id": 15,
    "name": "Albert Wagner Fire Safety",
    "logoUrl": "https://placekitten.com/60/70",
    "specialties": [
      "FIRE_SAFETY"
    ],
    "city": "Berlin"
  }
]