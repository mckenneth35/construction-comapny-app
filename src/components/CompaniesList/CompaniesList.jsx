import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
// import throttle from 'lodash.throttle';
// import debounce from 'lodash.debounce';
// import InfiniteScroll from 'react-infinite-scroll-component';
// import chroma from 'chroma-js';

import CompanySummary from '../CompanySummary';
import Company from '../../types/Company';

// import { COMPANIES_LIST_PAGING_SIZE } from '../../constants/pagingConfig';
import { SPECIALTY_TYPES, SPECIALTY_TYPES_NAME, DEFAULT_CHECKED_SPECIALTY_TYPES } from '../../constants/specialties';

const CompaniesList = ({
  companiesMap,
  companyIds,
  companiesLoading,
  onLoad,
}) => {
  useEffect(() => {
    onLoad();
  }, []);

  const [filteredName, setFilteredName] = useState('');
  const [checkedSpecialties, setCheckedSpecialties] = useState({ ...DEFAULT_CHECKED_SPECIALTY_TYPES });
  const initialRender = useRef(true);

  useEffect(() => {
    if (initialRender.current) {
      initialRender.current = false;
    } else {
      const criteria = {};
      if (filteredName) {
        criteria.name = filteredName;
      }

      const checkedSpecialtiesKeys = Object.keys(checkedSpecialties).filter((key) => checkedSpecialties[key]);
      if (checkedSpecialtiesKeys.length) {
        criteria.specialties = checkedSpecialtiesKeys;
      }
      onLoad(criteria);
    }
  }, [filteredName, checkedSpecialties]);

  return (
    <div>
      <div className='sticky-top top-panel'>
        <span id='Heading'>
          <h1>
            Companies
          </h1>
          <h3>
            {` (${companyIds.length} Result${companyIds.length > 1 ? 's' : ''})`}
          </h3>
        </span>

        <div className='name-search-div'>
          <input
            type='search'
            placeholder='Search by Company Name'
            onChange={(e) => setFilteredName(e.target.value)}
            id='Search-name'
            className='name-search-input'
          />
        </div>

        <span className='checkboxes-container'>
          {Object.keys(SPECIALTY_TYPES).map((code) => (
            <label className='checkbox-container' key={code}>
              <input
                type='checkbox'
                defaultChecked={checkedSpecialties[code]}
                onChange={(e) => setCheckedSpecialties({
                  ...checkedSpecialties,
                  [code]: !checkedSpecialties[code],
                })}
                id={code}
                className='specialty-checkbox'
              />
              <span className='checkbox-label'>
                {SPECIALTY_TYPES_NAME[code]}
              </span>
            </label>
          ))}
        </span>

        <hr />

        {/* table head */}
        <div className='companies-list-table-head'>
          <div className='company-summary-img' />
          <div className='company-summary-name'>Name</div>
          <div className='company-summary-specialties'>Specialties</div>
          <div className='company-summary-city'>City</div>
        </div>
        <hr />
      </div>

      {companiesLoading
        ? <h4>Loading</h4>
        : <div id='Companies-list' className='companies-list'>
          {companyIds.length === 0
            ? <h4>No Relevant Record Found</h4>
            : <>
              {companyIds.map((cId) => (
                <div key={cId}>
                  <CompanySummary company={companiesMap[cId]} />
                  <hr />
                </div>
              ))
              }
            </>
          }
        </div>
      }
    </div >
  );
};

// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
CompaniesList.propTypes = {
  companiesMap: PropTypes.objectOf(PropTypes.exact(Company)).isRequired,
  companyIds: PropTypes.arrayOf(PropTypes.number).isRequired,
  companiesLoading: PropTypes.bool.isRequired,
  onLoad: PropTypes.func.isRequired,
};

export default CompaniesList;
