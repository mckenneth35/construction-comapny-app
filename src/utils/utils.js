// eslint-disable-next-line no-sequences,no-return-assign,no-param-reassign
export const mapByKey = (arr, key) => (arr ? arr.reduce((map, element) => (map[element[key]] = element, map), {}) : {});
