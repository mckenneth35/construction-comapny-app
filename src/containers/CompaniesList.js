import { connect } from 'react-redux';

import CompaniesList from '../components/CompaniesList';
import { companiesMapSelector, companyIdsSelector, companiesLoadingSelector } from '../selectors/companiesSelectors';

const mapStateToProps = (state) => ({
  companiesMap: companiesMapSelector(state),
  companyIds: companyIdsSelector(state),
  companiesLoading: companiesLoadingSelector(state),
});

export default connect(mapStateToProps)(CompaniesList);
