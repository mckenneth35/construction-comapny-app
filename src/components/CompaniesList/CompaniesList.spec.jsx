import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import CompaniesList from './CompaniesList';
import { mapByKey } from '../../utils/utils'
import { mockCompanies } from '../../tests/data/Companies';

describe('component:CompaniesList', () => {
  it('Should render CompaniesList with 3 companies', () => {
    const wrapper = shallow(
      <CompaniesList
        companiesMap={mapByKey(mockCompanies, 'id')}
        companyIds={[1, 2, 3]}
        companiesLoading={true}
        onLoad={jest.fn()}
      />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
