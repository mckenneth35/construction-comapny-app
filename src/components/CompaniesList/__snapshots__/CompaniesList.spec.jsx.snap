// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:CompaniesList Should render CompaniesList with 3 companies 1`] = `
<div>
  <div
    className="sticky-top top-panel"
  >
    <span
      id="Heading"
    >
      <h1>
        Companies
      </h1>
      <h3>
         (3 Results)
      </h3>
    </span>
    <div
      className="name-search-div"
    >
      <input
        className="name-search-input"
        id="Search-name"
        onChange={[Function]}
        placeholder="Search by Company Name"
        type="search"
      />
    </div>
    <span
      className="checkboxes-container"
    >
      <label
        className="checkbox-container"
        key="EXCAVATION"
      >
        <input
          className="specialty-checkbox"
          defaultChecked={false}
          id="EXCAVATION"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="checkbox-label"
        >
          Excavation
        </span>
      </label>
      <label
        className="checkbox-container"
        key="PLUMBING"
      >
        <input
          className="specialty-checkbox"
          defaultChecked={false}
          id="PLUMBING"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="checkbox-label"
        >
          Plumbing
        </span>
      </label>
      <label
        className="checkbox-container"
        key="ELECTRICAL"
      >
        <input
          className="specialty-checkbox"
          defaultChecked={false}
          id="ELECTRICAL"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="checkbox-label"
        >
          Electrical
        </span>
      </label>
      <label
        className="checkbox-container"
        key="FIRE_SAFETY"
      >
        <input
          className="specialty-checkbox"
          defaultChecked={false}
          id="FIRE_SAFETY"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="checkbox-label"
        >
          Fire Safety
        </span>
      </label>
    </span>
    <hr />
    <div
      className="companies-list-table-head"
    >
      <div
        className="company-summary-img"
      />
      <div
        className="company-summary-name"
      >
        Name
      </div>
      <div
        className="company-summary-specialties"
      >
        Specialties
      </div>
      <div
        className="company-summary-city"
      >
        City
      </div>
    </div>
    <hr />
  </div>
  <h4>
    Loading
  </h4>
</div>
`;
